lib/aerospike/socket/ssl.rb

```diff
@@ -0,0 +1,83 @@
+require 'socket'
+require 'openssl'
+
+module Aerospike
+  module Socket
+    class SSL < Connection
+      attr_reader :context
+
+      def initialize(host, port, timeout, ssl_options)
+        @host, @port, @timeout = host, port, timeout
+        @tcp_socket = ::Socket.new(AF_INET, SOCK_STREAM, 0)
+        @context = create_context(ssl_options)
+        @socket = OpenSSL::SSL::SSLSocket.new(@tcp_socket, context)
+        @socket.sync_close = true
+      end
+
+      def connect!
+        @tcp_socket.connect(::Socket.sockaddr_in(port, host))
+        @socket.connect
+        verify_certificate!(@socket)
+        self
+      end
+
+      private
+
+      def create_context(options)
+        OpenSSL::SSL::SSLContext.new().tap do |ctx|
```

Comment:
Another team member suggested, that it might be best if we just let the user pass in a pre-configured `SSLContext` instance, instead of creating one ourself and exposing a subset of it's config options through our own settings. I like that idea, as it gives the user full flexibility to configure the context as needed without us having to expose every single config option through our own settings. On the other hand it does require a bit more work on the user's part for a standard setup and means we can no longer set sensible defaults, like restricting the connection to TLS v1.2+.

So, how about we allow the user to pass in a pre-configured context optionally, and create one if not. E.g. just add another `:context` option to the `ssl_options` hash. If not provided, we continue to create a new context as you have already implemented. But in that case it would be ok to not expose more advanced options such as setting the desired cipher suites, etc.

What do you think?

Reply:
@jhecking that's a great idea, to have both options. By allowing people to pass a custom context we'd potentially avoid a bunch of future requests asking for less common options. And as a lazy Ruby developer I like the simple option of just quickly enable SSL with sane defaults

In your experience, which options are the most commonly used by AS users? Anything besides the ones in this PR?

Reply:
> In your experience, which options are the most commonly used by AS users? Anything besides the ones in this PR?

I think we should keep it to the minimum needed to get a working connection to a standard server deployment, with sensible defaults for everything else. I would support:
* `:enable`
* `:ca_file`, `:ca_path`
* `:cert_file`
* `:key_file`, `:key_file_pass_phrase`

You can remove the `:cipher_suite` and `:protocols` options and just default `SSLContext#min_version=` to `OpenSSL::SSL::TLS1_2_VERSION`.